### PR TITLE
gen_vimdoc.py: add more structure to mpack result

### PR DESCRIPTION
**Pull request by smolck**
See TODO in https://github.com/neovim/neovim/commit/46bde66147ead1bb234932a48e2c56fc4b698ec0.

Definitely isn't perfect, so any feedback is appreciated.

EDIT: There's likely to be some unnecessary code due to my re-purposing of the functions already in gen_vimdoc.py, so please let me know if you see any such code or functions.

### CODE REVIEW

File: scripts/gen_vimdoc.py [hunk 1]
```diff
@@ -365,6 +395,86 @@ def render_node(n, text, prefix='', indent='', width=62):
...
+def para_as_map(parent, indent='', width=62):
+    """Parses XML <para> as a map
+
+    The map returned may or may not contain
+    all of the following keys:
+        'text': Text in the para tag
+        'params': Information from <parameterlist>'s
+        'return': TODO
+        'seealso': TODO
+        'xrefs': TODO
```
Comment (smolck) [inline scripts/gen_vimdoc.py]: Not so sure on what to put for the docstrings here. Also, is `xrefs` necessary?

File: scripts/gen_vimdoc.py [hunk 2]
```diff
@@ -627,6 +868,7 @@ def gen_docs(config):
         if p.returncode:
             sys.exit(p.returncode)
 
+        mpacks = {}
```
Comment (smolck) [inline scripts/gen_vimdoc.py]: Maybe need a different name for this?

File: scripts/gen_vimdoc.py [hunk 3]
```diff
@@ -443,13 +563,134 @@ def parse_parblock(parent, prefix='', width=62, indent=''):
 # }}}
 
 
+def parse_source_xml_for_mpack(filename, mode):
```
Comment (smolck) [inline scripts/gen_vimdoc.py]: Different name?

File: scripts/gen_vimdoc.py [hunk 4]
```diff
@@ -262,6 +262,36 @@ def has_nonexcluded_params(nodes):
...
+
+        items.append((name.strip(), node))
+
+    out = {}  # Map of name:desc
+    for name, node in items:
+        desc = ''
+        desc_node = get_child(node, 'parameterdescription')
+        if desc_node:
+            desc = parse_parblock(desc_node, width=width,
+                                  indent=(' ' * len(name)))
```
Comment (smolck) [inline scripts/gen_vimdoc.py]: Should probably remove this `indent=` part?

File: scripts/gen_vimdoc.py (hunk 1 above)
Comment (justinmk) [reply scripts/gen_vimdoc.py]: not sure. xrefs might be useful for docstrings that "link" to each other, but currently it's just there "for fun".

**Comment by justinmk** [comment]
> There's likely to be some unnecessary code due to my re-purposing of the functions already in gen_vimdoc.py, so please let me know if you see any such code or functions

I pushed a commit which eliminates the extra `parse_source_xml()` variant. Somewhat hacky in that I added a new paramter `fmt_vimdoc`, but this is preferable to having many big-yet-similar functions.

Maybe the same treatment could be applied to the other cases?

File: scripts/gen_vimdoc.py [hunk 5]
```diff
@@ -365,6 +396,86 @@ def render_node(n, text, prefix='', indent='', width=62):
     return text
 
 
+def para_as_map(parent, indent='', width=62):
+    """Parses XML <para> as a map
+
+    The map returned may or may not contain
+    all of the following keys:
```
Comment (justinmk) [inline scripts/gen_vimdoc.py]: Makes things a lot nicer if callers can always assume keys exist, no?

**Comment by justinmk** [comment]
I regressed this a bit. Will fix later.

File: scripts/gen_vimdoc.py [hunk 6]
```diff
@@ -563,21 +564,23 @@ def parse_parblock(parent, prefix='', width=62, indent=''):
...
 
     Returns two maps:
-      1. API functions
-      2. Deprecated API functions
+      1. Functions
+      2. Deprecated functions
 
-    Caller decides what to do with the deprecated documentation.
+    The `fmt_vimhelp` parameter controls some special cases for use by
+    fmt_doxygen_xml_as_vimhelp(). (TODO: ugly :)
```
Comment (smolck) [inline scripts/gen_vimdoc.py]: What do you think of building the Vim help text in `fmt_doxygen_xml_as_vimhelp()` directly from the function maps returned by `extract_from_xml`? That way there’s no need for the fmt_vimhelp and desc_node stuff. Only thing is that I’m unsure how much more work that’d be.

Comment (justinmk) [reply scripts/gen_vimdoc.py]: That would be ideal.

**Comment by justinmk** [comment]
@smolck  latest version produces pretty good `*.mpack` results. There are some formatting regressions for the Vim `:help` results, but I'm close to fixing that. Meanwhile I'll merge this. 

Remaining question for the "unformatted" (`*.mpack`) forms is how to represent lists and line breaks. Currently line breaks (and indent) are included in the `parameters_doc` map. It's easy enough for API clients to just strip linebreak-preceded whitespace, but that will break `<pre>` formatted blocks (and nested lists).

Here's an example from `api.mpack`:

```
"nvim_replace_termcodes": {
    "annotations": [],
    "doc": [
        "Replaces terminal codes and |keycodes| (<CR>, <Esc>, ...) in a\nstring with the internal representation."
    ],
    "parameters": [
        [
            "String",
            "str"
        ],
        [                      
            "Boolean",                                                   
            "from_part"                                   
        ],                                                           
        [
            "Boolean",                          
            "do_lt"           
        ],        
        [                                                             
            "Boolean",
            "special"         
        ]                
    ],           
    "parameters_doc": {                                                                                           
        "do_lt": "Also translate <lt>. Ignored if `special` is false.",                 
        "from_part": "Legacy Vim parameter. Usually true.",                 
        "special": "Replace |keycodes|, e.g. <CR> becomes a \"\\n\" char.",                                                  
        "str": "String to be converted."                                                                                                      
    },           
    "return": [],      
    "seealso": [
        "    replace_termcodes",                                                        
        "    cpoptions"
    ],                                                               
    "signature": "nvim_replace_termcodes({str}, {from_part}, {do_lt}, {special})"
}, 
```

**Comment by smolck** [comment]
LGTM! Thanks for all the help with this!

**Comment by justinmk** [comment]
@smolck Note that the mpack structure changed. Here's a 2-liner to preview it in human-readable form:

```
:new | put=json_encode(msgpackparse(readfile('runtime/doc/api.mpack')))
:%!python -m json.tool
```

**Comment by blueyed** [comment]
btw: running `scripts/gen_vimdoc.py` changes `runtime/doc/api.txt` and `runtime/doc/lua.txt`.  Is that intentional?

**Comment by justinmk** [comment]
@blueyed that's what I referred to here:

> There are some formatting regressions for the Vim `:help` results, but I'm close to fixing that.

**Pull request merged**